File: KNI.Engine/TankGame/KFramework.MonoGame/KFramework.MonoGame/Timer/KTimeOutGenerator.cs
namespace KFramework.MonoGame
{
    public class KTimeOutGenerator
    {
        float fLastUpdateTime = 0;
        float fInternalTime = 0;
        public static KTimeOutGenerator New(float fInternalTime)
        {
            var temp = new KTimeOutGenerator();
            temp.Init(fInternalTime);
            return temp;
        }

        public void Init(float fInternalTime = 1.0f)
        {
            this.fInternalTime = fInternalTime;
            this.Reset();
        }

        public void Reset()
        {
            this.fLastUpdateTime = KTime.time;
        }

        public bool orTimeOut()
        {
            if ((KTime.time - fLastUpdateTime) > fInternalTime)
            {
                this.Reset();
                return true;
            }

            return false;
        }

        public bool orTimeOutWithSpecialTime(float fInternalTime)
        {
            if (KTime.time - fLastUpdateTime > fInternalTime)
            {
                this.Reset();
                return true;
            }

            return false;
        }
    }
}

File: PixiJS.Engine/TankGame/src/KFramework.PixiJS/Timer/KTimeOutGenerator.ts
import { KTime } from "../GameEngine/KTime";

export class KTimeOutGenerator
{
    private fTime:number = 0;
    private fInternalTime:number = 0;

    public static New(fInternalTime:number):KTimeOutGenerator
    {
        var temp = new KTimeOutGenerator();
        temp.Init(fInternalTime);
        return temp;
    }

    public Init(fInternalTime:number = 1.0):void
    {
        this.fInternalTime = fInternalTime;
        this.Reset();
    }
    
    public Reset():void
    {
        this.fTime = 0;
    }

    public orTimeOut():boolean
    {
        this.fTime += KTime.deltaTime;
        if (this.fTime > this.fInternalTime)
        {
            this.Reset();
            return true;
        }

        return false;
    }
    
    public orTimeOutWithSpecialTime(fInternalTime:number):boolean
    {
        this.fTime += KTime.deltaTime;
        if (this.fTime > fInternalTime)
        {
            this.Reset();
            return true;
        }

        return false;
    }
    
}
